packages/discord-request/src/queue.ts

```diff
@@ -0,0 +1,238 @@
+		}
+
+		this.manager.globalRequestCounter -= 1;
+
+		// Setup the timeout signal
+		const signal = new AbortController();
+		const abortTimeout = setTimeout(() => {
+			signal.abort();
+		}, this.manager.config.timeout);
+		const clearAbort = () => {
+			clearTimeout(abortTimeout);
+		};
+
+		this.#shutdownSignal?.addEventListener("signal", clearAbort);
+
+		// Callbacks
+		this.manager.onRequest?.(route, resource, init, retries);
+
+		// Send the request
+		const request = new Request(resource, init);
+		let response: Response;
+
+		try {
+			response = await fetch(request, { signal: signal.signal });
+		} catch (error: unknown) {
+			// Handle timeout aborts
+			if (
+				error instanceof Error &&
+				error.name === "AbortError" &&
+				retries < this.manager.config.retries
+			) {
+				return await this.#process(route, resource, init, retries + 1);
+			}
+
+			throw error;
+		} finally {
+			this.#shutdownSignal?.removeEventListener("signal", clearAbort);
+			clearAbort();
+		}
+
+		// Parse Rate Limit information
+		let retryAfter = 0;
+
+		const global = Boolean(response.headers.get("X-RateLimit-Global"));
+		const key = response.headers.get("X-RateLimit-Bucket");
+
+		const retry = response.headers.get("Retry-After");
+		const reset = response.headers.get("X-RateLimit-Reset-After");
+		const limit = response.headers.get("X-RateLimit-Limit");
+		const remaining = response.headers.get("X-RateLimit-Remaining");
+
+		this.reset = reset
+			? 1000 * Number(reset) + Date.now() + OFFSET
+			: Date.now();
+		this.limit = limit ? Number(limit) : Infinity;
+		this.remaining = remaining ? Number(remaining) : 1;
+
+		if (retry != null) {
+			retryAfter = Number(retry) * 1000 + OFFSET;
+		}
+
+		// Handle global rate limits
+		if (global && retryAfter > 0) {
+			this.manager.globalRequestCounter = 0;
+			this.manager.globalReset = Date.now() + retryAfter;
+		}
+
+		// Update bucket hashes as needed
+		const identifier = getRouteKey(init.method as RequestMethod, route);
+
+		if (key != null && key !== this.id) {
+			log(`Received new bucket hash. ${this.id} -> ${key}`);
+
+			this.manager.buckets.set(identifier, {
+				key,
+				lastRequest: Date.now(),
+			});
+		} else if (key != null && this.manager.buckets.has(identifier)) {
+			this.manager.buckets.set(identifier, {
+				key,
+				lastRequest: Date.now(),
+			});
+		}
+
+		// Successful Requests
+		if (response.ok) {
+			return response.json();
+		}
+
+		// Rate Limited Requests
+		if (response.status === 429) {
+			const rateLimitData = {
+				retryAfter,
+				bucket: this.id,
+				url: resource,
+				limit: this.limit,
+				route: route.path,
+				identifier: route.identifier,
+				global: global,
+				method: init.method as RequestMethod,
+			};
+
+			this.manager.onRateLimit?.(rateLimitData);
+
+			log(`Encountered 429 rate limit. ${JSON.stringify(rateLimitData)}.`);
+
+			// If caused by a sublimit, wait it out here so other requests on the route can be handled
```

Comment:
This isn't true at all, it stalls the entire queue until the sublimit passes anyways, but does still need to happen here

Reply:
Whoops yeah, bad leftover comment.